---
title: "VCFugit"
author: "Jerod Parsons"
date: "August 1, 2016"
output: 
  html_document:
    code_folding: hide
runtime: shiny
---
## Variant Annotation of a provided VCF file for Tempus (Code Demo)

This report is written in markdown because interactive HTML output is often more useful than static text for scientific reporting.  Although this example report is a bit lacking in practical applications, it is still relevant to be able to filter and sort the datatable output interactively.  This application could also do some more complex and interactive data filtering steps prior to the data table output if required.  To look at the code, you can simply click on the 'code' button to expand code chunks.

```{r Setup}
knitr::opts_chunk$set(tidy=TRUE) 
```
```{r PackageInstallation}
# Package installation section: Install required packages
biocpackages<-c("VariantAnnotation","GenomicFeatures","biomaRt",
                "BSgenome.Hsapiens.UCSC.hg19","TxDb.Hsapiens.UCSC.hg19.knownGene")
cranpackages<-c("data.table","shiny","gdata")
if(sum(!biocpackages%in%installed.packages())>0){
  source("https://bioconductor.org/biocLite.R")
  biocLite(biocpackages[!biocpackages%in%installed.packages()])}
if(sum(!cranpackages%in%installed.packages())>0){
  install.packages(cranpackages[!cranpackages%in%installed.packages()],repos='http://cran.us.r-project.org')
}
```
Because these bioconductor packages most easily use the hg19 genome and this vcf was mapped to GRCh37, some genomic coordinates 
do not match.  NA's in the table below show how many variants were unable to be located on the Hg19 reference.
```{r DefineDataset}
buildvcftable<-function(vcffile,genome){
  if(is.null(vcffile)){return(0)}

suppressPackageStartupMessages({  library("VariantAnnotation")
  library("GenomicFeatures")
  library("biomaRt")
  library("BSgenome.Hsapiens.UCSC.hg19")
  library("data.table")
})

  vcf<-readVcf(vcffile,genome = genome)
# Consider speeding up the vcf read by using ScanVcfParam 
# Although that may cause issues with predictCoding and locateVariants
  switch(genome,
         'hg19'=library("TxDb.Hsapiens.UCSC.hg19.knownGene"),
         stop("Unsupported genome ",genome))
# Currently only supporting HG19 because unfortunately that's what all of the TxDbs support.
# It comes at the cost of interpretation of the mitochondria and other non-chromosomal sequences in grch37/38 data.
  
# Determine Type of variation 
  seqlevelsStyle(vcf)<-"UCSC"
  all <- suppressWarnings(suppressMessages(locateVariants(vcf, TxDb.Hsapiens.UCSC.hg19.knownGene, AllVariants()))) # Genomic location of variant
  mcols(all)$LOCATION<-gdata::reorder.factor(mcols(all)$LOCATION,new.order=c("intergenic","intron","threeUTR","fiveUTR","promoter","spliceSite","coding")) 
# Re-order the levels of the factor in order of least to most detrimental, so i can just use 'max' to return the most deleterious
  variants<-split(mcols(all)$LOCATION,mcols(all)$QUERYID)
  maxfactor<-function(input){
    # Input:  Factor of locations.  Factor must be ordered such that the last level is 'worst'
    # Return:  The highest level present in the factor
  levels(input)[max(as.numeric(input))]
}
varl <- lapply(variants,maxfactor) # A list of variants and the worst of locations
varl <- data.frame(pos=names(varl),var=unlist(varl))
cdiff <- predictCoding(vcf,TxDb.Hsapiens.UCSC.hg19.knownGene,Hsapiens) 
# Predict nature of those various coding variants
conseq <- split(mcols(cdiff)$CONSEQUENCE,mcols(cdiff)$QUERYID)
mcols(cdiff)$CONSEQUENCE<-gdata::reorder.factor(mcols(cdiff)$CONSEQUENCE,new.order=c("frameshift","nonsense","nonsynomyous","synonymous"))
conseql <- lapply(conseq,maxfactor)
conseql <- data.frame(pos=names(conseql),var=unlist(conseql))
# Now merge together the locations & consequences
Variants <- data.frame(pos=1:length(vcf))
varl$var <- factor(varl$var,levels=c(levels(varl$var),"frameshift","nonsense","nonsynonymous","synonymous")) # Add the consequences into the same factor list
if(length(conseql$var) == sum(varl$var=="coding")){
  varl$var[varl$var=="coding"]<-conseql$var
}# A quick test that nothing went wrong, and an overwrite of "coding" to its various subtypes
Variants <- merge(Variants,varl,by="pos",all="TRUE")
# Depth of coverage @ position
outdf<-data.frame(info(vcf)[c("TR","TC")])
# Number of reads supporting variant
outdf$TR <- unlist(lapply(outdf$TR, `[[`, 1)) # The TR field is a list of integerlists,
# we just want the 1st element.  This causes technically incorrect results 
# in 15/11765 test cases.

# Percent of reads supporting variant
outdf$ReadPct <- round(outdf$TR/outdf$TC,digits = 3)*100
# Also relevant information: Filter Status:
Filter=mcols(vcf)["FILTER"]
cpr<-strsplit(rownames(outdf),":")
ra<-strsplit(unlist(lapply(cpr,'[[',2)),"_")
vcftable<-data.table(Chromosome=lapply(cpr,'[[',1),
                     Position=unlist(lapply(strsplit(unlist(lapply(cpr,'[[',2)),"_"),'[[',1)),
                     reference=unlist(lapply(strsplit(unlist(lapply(ra,'[[',2)),"/"),'[[',1)),
                     alt=unlist(lapply(strsplit(unlist(lapply(ra,'[[',2)),split="/"),'[[',2)),
                     VarType=Variants$var,Depth=outdf$TC,Supporting_Reads=outdf$TR,
                     Percent_Support=outdf$ReadPct,Filter=Filter$FILTER)
return(vcftable)
}
vcffile<-"test_vcf_data.vcf"
if(!file.exists(gsub(vcffile,".vcf",".rData"))){
vcftable<-buildvcftable(vcffile=vcffile,genome="hg19")
save(vcftable,file=gsub(vcffile,".vcf",".rData"))} # some rudimentary caching to keep this from running forever.
if(file.exists(gsub(vcffile,".vcf",".rData"))){load(gsub(vcffile,".vcf",".rData"))}
summary((vcftable$VarType))
```
```{R DisplayOutput}
textInput('filename',label="filename to save to",value="test.csv")
actionButton('go','Save!')
observeEvent(input$go,
  write.csv(x = vcftable,file=input$filename,quote=FALSE,row.names = FALSE))
renderDataTable(vcftable)
```

## Output Description:  
1. Chromosome - Number of the chromosome the variant is on.  
2. Position - Genome coordinates of the variant  
3. reference - Sequence at this position in the reference  
4. alt - Sequence at this position in the variant.  
5. VarType - Location and effect of variant  
  + fiveUTR: 5' Untranslated region of a gene  
  + threeUTR: 3' Untranslated region  
  + intron: within an intron  
  + promoter: within a promoter region  
  + intergenic: not in or around a coding region  
  + spliceSite: at a splice junction  
  + synonymous: Coding variant with no effect on protein sequence  
  + nonsynonymous: Coding variant with an effect on protein sequence  
  + frameshift: Coding variant that causes a shift in protein reading frame  
  + nonsense: Coding variant that introduces a new stop codon  
6. Depth - Number of reads covering this position  
7. Supporting_Reads - Number of reads at this position with this variant  
8. Percent_Support - Proportion of reads that support this variant  
9. Filter - Confidence measure of variant; do we have a reason to expect the sequence to be a bias?  
  + Pass: This variant passes all filter criteria  
  + AlleleBias: This variant is supported by an unexpected proportion of reads (expected: 50%, 100%)  
  + HapScore: More than an expected number of haplotypes are called  
  + SC: variant is in a low-complexity sequence  
  + strandBias: Variant shows evidence of strand bias  